src/realm/query_expression.cpp

```diff
@@ -217,6 +218,104 @@ std::vector<ObjKey> LinkMap::get_origin_ndxs(ObjKey key, size_t column) const
     return ret;
 }
 
+Columns<Dictionary>& Columns<Dictionary>::key(const Mixed& key_value)
+{
+    if (m_key_type != type_Mixed && key_value.get_type() != m_key_type) {
+        throw LogicError(LogicError::collection_type_mismatch);
+    }
+
+    m_key = key_value;
+    if (!key_value.is_null()) {
+        auto hash = key_value.hash();
+        m_objkey = ObjKey(int64_t(hash & 0x7FFFFFFFFFFFFFFF));
+    }
+    return *this;
+}
+
+void Columns<Dictionary>::set_cluster(const Cluster* cluster)
+{
+    m_array_ptr = nullptr;
+    m_leaf_ptr = nullptr;
+    if (links_exist()) {
+        m_link_map.set_cluster(cluster);
+    }
+    else {
+        // Create new Leaf
+        m_array_ptr = LeafPtr(new (&m_leaf_cache_storage) ArrayInteger(m_link_map.get_base_table()->get_alloc()));
+        cluster->init_leaf(m_column_key, m_array_ptr.get());
+        m_leaf_ptr = m_array_ptr.get();
+    }
+}
+
+void Columns<Dictionary>::evaluate(size_t index, ValueBase& destination)
+{
+    Value<Mixed>& d = static_cast<Value<Mixed>&>(destination);
+
+    if (links_exist()) {
+        std::vector<ObjKey> links = m_link_map.get_links(index);
+        auto sz = links.size();
+
+        std::vector<Mixed> values;
+        for (size_t t = 0; t < sz; t++) {
+            const Obj obj = m_link_map.get_target_table()->get_object(links[t]);
+            auto dict = obj.get_dictionary(m_column_key);
+            if (m_key.is_null()) {
+                // Insert all values
+                dict.for_all_values([&values](const Mixed& value) {
+                    values.emplace_back(value);
+                });
+            }
+            else {
+                auto opt_val = dict.try_get(m_key);
+                if (opt_val)
+                    values.emplace_back(*opt_val);
```

Comment:
I think you might want to insert null if try_get fails? at least that seems consistent with the behaviour below with no links...

Reply:
You're right. A test for equality against null should also be possible over links.